src/main/java/spoon/support/adaption/TypeAdaptor.java

```diff
@@ -0,0 +1,598 @@
+		if (subMethod.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			return getOldClassTypingContext().isOverriding(subMethod, superMethod);
+		}
+		if (subMethod.getParameters().size() != superMethod.getParameters().size()) {
+			return false;
+		}
+		if (!subMethod.getSimpleName().equals(superMethod.getSimpleName())) {
+			return false;
+		}
+
+		if (subMethod.isStatic() || superMethod.isStatic()) {
+			return false;
+		}
+
+		CtType<?> subDeclaringType = subMethod.getDeclaringType();
+		CtType<?> superDeclaringType = superMethod.getDeclaringType();
+
+		if (!isSubtype(subDeclaringType, superDeclaringType.getReference())) {
+			return false;
+		}
+
+		CtMethod<?> adapted = new TypeAdaptor(subMethod.getDeclaringType())
+			.adaptMethod(superMethod);
+
+		for (int i = 0; i < subMethod.getParameters().size(); i++) {
+			CtParameter<?> subParam = subMethod.getParameters().get(i);
+			CtParameter<?> superParam = adapted.getParameters().get(i);
+
+			if (!subParam.getType().getTypeErasure().equals(superParam.getType().getTypeErasure())) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Adapts a type from a supertype to the context of this adaptor. In essence, this method builds
+	 * the inheritance hierarchy from its context to the super reference (with some smarts to figure
+	 * out what to do if superRef is actually a type parameter) and then walks backwards along that
+	 * chain until it finds a type variable in the adaptor's context or a terminating type.
+	 * <p>
+	 * For example:
+	 * <pre>{@code
+	 *   interface Top<T, S> {}
+	 *   interface Middle<Q> extends Top<Q, String> {}
+	 *   interface Bottom<R> extends Middle<R> {}
+	 * }</pre>
+	 * If you adapt {@code T} from {@code Top} to {@code Middle} you get {@code Q}. If you adapt
+	 * {@code T} from {@code Top} to {@code Bottom} you get {@code R}.
+	 * <br>If you adapt {@code S} from {@code Top} to {@code Middle}/{@code Bottom} you get {@code
+	 * String}.
+	 *
+	 * <br>If the input reference is a formal type parameter declared on a method, adaption is only
+	 * possible if this adaptor was created using {@link #TypeAdaptor(CtMethod)}. Otherwise, there is
+	 * no reference method to adapt to and the input will be returned unchanged.
+	 * <br>If that constructor was used, this method will return the corresponding type parameter
+	 * declared on the context method of this adaptor.
+	 *
+	 * @param superRef the super type to adapt
+	 * @return the adapted type
+	 */
+	@SuppressWarnings("removal")
+	public CtTypeReference<?> adaptType(CtTypeReference<?> superRef) {
+		if (superRef.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			if (startMethod != null) {
+				return new MethodTypingContext()
+					.setClassTypingContext(getOldClassTypingContext())
+					.setMethod(startMethod)
+					.adaptType(superRef);
+			}
+			return getOldClassTypingContext().adaptType(superRef);
+		}
```

Comment:
For all methods that have a legacy alternative, I think it's a good thing to minimize the noise. So, something like this:

```suggestion
	public CtTypeReference<?> adaptType(CtTypeReference<?> superRef) {
		if (useLegacyTypeAdaption(superRef)) {
			return legacyAdaptType(superRef);
		}
```

And then have a helper `useLegacyTypeAdaption` with the `SuppressWarnings` on it. Even if the code in the legacy implementation is a one-liner, I think it helps to separate it into a method of its own just for clarity.

Reply:
I would also suggest renaming the flag to `useLegacyTypeAdaption`.